File: SiriusQuality/SiriusQuality-ET/src/simplace/SiriusQuality_ET/NetRadiation.java
package net.simplace.sim.components.SiriusQuality-ET;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class NetRadiation extends FWSimComponent
{
    private FWSimVariable<Double> albedoCoefficient;
    private FWSimVariable<Double> maxTair;
    private FWSimVariable<Double> minTair;
    private FWSimVariable<Double> vaporPressure;
    private FWSimVariable<Integer> ih;
    private FWSimVariable<Double> extraSolarRadiation;
    private FWSimVariable<Double> solarRadiation;
    private FWSimVariable<Double> tau;
    private FWSimVariable<Double> elevation;
    private FWSimVariable<Double> stefanBoltzman;
    private FWSimVariable<Double> albedoCoefficientCan;
    private FWSimVariable<Double> netRadiation;
    private FWSimVariable<Double> netOutGoingLongWaveRadiation;

    public NetRadiation(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public NetRadiation(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("albedoCoefficient", "albedo Coefficient", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.23, this));
        addVariable(FWSimVariable.createSimVariable("maxTair", "maximum air Temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", 30, 45, 7.2, this));
        addVariable(FWSimVariable.createSimVariable("minTair", "minimum air temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", 30, 45, 0.7, this));
        addVariable(FWSimVariable.createSimVariable("vaporPressure", "vapor Pressure", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"hPa", 0, 1000, 6.1, this));
        addVariable(FWSimVariable.createSimVariable("ih", "hour of the day if the component is hourly, -999 if the component is daily", DATA_TYPE.INT, CONTENT_TYPE.state,"", 999, 24, 999, this));
        addVariable(FWSimVariable.createSimVariable("extraSolarRadiation", "extra Solar Radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"MJ m2 d-1", 0, 1000, 11.7, this));
        addVariable(FWSimVariable.createSimVariable("solarRadiation", "solar Radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"MJ m-2 d-1", 0, 1000, 3, this));
        addVariable(FWSimVariable.createSimVariable("tau", "plant cover factor", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 100, 0.9983, this));
        addVariable(FWSimVariable.createSimVariable("elevation", "elevation", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"m", 500, 10000, 0, this));
        addVariable(FWSimVariable.createSimVariable("stefanBoltzman", "stefan Boltzman constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 4.903E-09, this));
        addVariable(FWSimVariable.createSimVariable("albedoCoefficientCan", "albedo Coefficient", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.23, this));
        addVariable(FWSimVariable.createSimVariable("netRadiation", "net radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.out,"MJ m-2 d-1", 0, 5000, null, this));
        addVariable(FWSimVariable.createSimVariable("netOutGoingLongWaveRadiation", "net OutGoing Long Wave Radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.out,"g m-2 d-1", 0, 5000, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        double t_albedoCoefficient = albedoCoefficient.getValue();
        double t_maxTair = maxTair.getValue();
        double t_minTair = minTair.getValue();
        double t_vaporPressure = vaporPressure.getValue();
        Integer t_ih = ih.getValue();
        double t_extraSolarRadiation = extraSolarRadiation.getValue();
        double t_solarRadiation = solarRadiation.getValue();
        double t_tau = tau.getValue();
        double t_elevation = elevation.getValue();
        double t_stefanBoltzman = stefanBoltzman.getValue();
        double t_albedoCoefficientCan = albedoCoefficientCan.getValue();
        double t_netRadiation = netRadiation.getDefault();
        double t_netOutGoingLongWaveRadiation = netOutGoingLongWaveRadiation.getDefault();
        double Nsr;
        double clearSkySolarRadiation;
        double averageT;
        double surfaceEmissivity;
        double cloudCoverFactor;
        double Nolr;
        double cov;
        if (t_ih == -999)
        {
            Nsr = t_solarRadiation * (1 - (t_albedoCoefficientCan * t_tau + (t_albedoCoefficient * (1.00d - t_tau))));
        }
        else
        {
            cov = (double)(1);
            if (t_solarRadiation > 0.01d)
            {
                if (t_ih <= 7)
                {
                    cov = 0.30d;
                }
                else if ( t_ih > 7 && t_ih < 11)
                {
                    cov = 0.30d - (0.09d / 3.00d * (t_ih - 7.00d));
                }
                else if ( t_ih == 11)
                {
                    cov = 0.21d;
                }
                else if ( t_ih > 11 && t_ih < 15)
                {
                    cov = 0.21d + (0.09d / 3.00d * (t_ih - 11.00d));
                }
                else
                {
                    cov = 0.30d;
                }
            }
            Nsr = (1 - cov) * t_solarRadiation;
        }
        clearSkySolarRadiation = (0.750d + (2 * Math.pow(10.00d, -5) * t_elevation)) * t_extraSolarRadiation;
        averageT = (Math.pow(t_maxTair + 273.160d, 4) + Math.pow(t_minTair + 273.160d, 4)) / 2.00d;
        surfaceEmissivity = 0.340d - (0.140d * Math.sqrt(t_vaporPressure / 10.00d));
        cloudCoverFactor = 1.350d * (t_solarRadiation / clearSkySolarRadiation) - 0.350d;
        Nolr = t_stefanBoltzman * averageT * surfaceEmissivity * cloudCoverFactor;
        if (t_ih != -999)
        {
            Nolr = Nolr / 24.00d;
        }
        t_netRadiation = Nsr - Nolr;
        t_netOutGoingLongWaveRadiation = Nolr;
        netRadiation.setValue(t_netRadiation, this);
        netOutGoingLongWaveRadiation.setValue(t_netOutGoingLongWaveRadiation, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new NetRadiation(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}

File: SiriusQuality/SiriusQuality-ET/src/openalea/SiriusQuality_EnergyBalance/netradiation.py
# coding: utf8
from copy import copy
from array import array
from math import *
from typing import *
from datetime import datetime

import numpy

#%%CyML Model Begin%%
def model_netradiation(minTair:float,
         maxTair:float,
         ih:int,
         tau:float,
         albedoCoefficientCan:float,
         elevation:float,
         stefanBoltzman:float,
         vaporPressure:float,
         solarRadiation:float,
         extraSolarRadiation:float,
         albedoCoefficient:float):
    """
     - Name: NetRadiation -Version: 1.0, -Time step: 1
     - Description:
                 * Title: NetRadiation Model
                 * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
                 * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
                 * Institution: New Zealand Institute for Crop and Food Research Ltd.,
     New Zealand Institute for Crop and Food Research Ltd.,
     New Zealand Institute for Crop and Food Research Ltd.,
     New Zealand Institute for Crop and Food Research Ltd.
     
                 * ExtendedDescription: It is calculated at the surface of the canopy and is givenby the difference between incoming and outgoing radiation of both short
     and long wavelength radiation
                 * ShortDescription: It refers as difference between incoming and outgoing radiation of both short
     and long wavelength radiation
     - inputs:
                 * name: minTair
                               ** description : minimum air temperature
                               ** inputtype : variable
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** max : 45
                               ** min : 30
                               ** default : 0.7
                               ** unit : degC
                 * name: maxTair
                               ** description : maximum air Temperature
                               ** inputtype : variable
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** max : 45
                               ** min : 30
                               ** default : 7.2
                               ** unit : degC
                 * name: ih
                               ** description : hour of the day if the component is hourly, -999 if the component is daily
                               ** inputtype : variable
                               ** parametercategory : state
                               ** datatype : INT
                               ** max : 24
                               ** min : 999
                               ** default : 999
                               ** unit : 
                 * name: tau
                               ** description : plant cover factor
                               ** inputtype : parameter
                               ** parametercategory : species
                               ** datatype : DOUBLE
                               ** max : 100
                               ** min : 0
                               ** default : 0.9983
                               ** unit : 
                 * name: albedoCoefficientCan
                               ** description : albedo Coefficient
                               ** inputtype : parameter
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** max : 1
                               ** min : 0
                               ** default : 0.23
                               ** unit : 
                 * name: elevation
                               ** description : elevation
                               ** inputtype : parameter
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** max : 10000
                               ** min : 500
                               ** default : 0
                               ** unit : m
                 * name: stefanBoltzman
                               ** description : stefan Boltzman constant
                               ** inputtype : parameter
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** max : 1
                               ** min : 0
                               ** default : 4.903E-09
                               ** unit : 
                 * name: vaporPressure
                               ** description : vapor Pressure
                               ** inputtype : variable
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** max : 1000
                               ** min : 0
                               ** default : 6.1
                               ** unit : hPa
                 * name: solarRadiation
                               ** description : solar Radiation
                               ** inputtype : variable
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** max : 1000
                               ** min : 0
                               ** default : 3
                               ** unit : MJ m-2 d-1
                 * name: extraSolarRadiation
                               ** description : extra Solar Radiation
                               ** inputtype : variable
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** max : 1000
                               ** min : 0
                               ** default : 11.7
                               ** unit : MJ m2 d-1
                 * name: albedoCoefficient
                               ** description : albedo Coefficient
                               ** inputtype : parameter
                               ** parametercategory : constant
                               ** datatype : DOUBLE
                               ** max : 1
                               ** min : 0
                               ** default : 0.23
                               ** unit : 
     - outputs:
                 * name: netRadiation
                               ** description : net radiation
                               ** datatype : DOUBLE
                               ** variablecategory : auxiliary
                               ** max : 5000
                               ** min : 0
                               ** unit : MJ m-2 d-1
                 * name: netOutGoingLongWaveRadiation
                               ** description : net OutGoing Long Wave Radiation
                               ** datatype : DOUBLE
                               ** variablecategory : auxiliary
                               ** max : 5000
                               ** min : 0
                               ** unit : g m-2 d-1
    """

    netRadiation:float
    netOutGoingLongWaveRadiation:float
    Nsr:float
    clearSkySolarRadiation:float
    averageT:float
    surfaceEmissivity:float
    cloudCoverFactor:float
    Nolr:float
    cov:float
    if ih == -999:
        Nsr = solarRadiation * (1 - (albedoCoefficientCan * tau + (albedoCoefficient * (1.00 - tau))))
    else:
        cov = float(1)
        if solarRadiation > 0.01:
            if ih <= 7:
                cov = 0.30
            elif ih > 7 and ih < 11:
                cov = 0.30 - (0.09 / 3.00 * (ih - 7.00))
            elif ih == 11:
                cov = 0.21
            elif ih > 11 and ih < 15:
                cov = 0.21 + (0.09 / 3.00 * (ih - 11.00))
            else:
                cov = 0.30
        Nsr = (1 - cov) * solarRadiation
    clearSkySolarRadiation = (0.750 + (2 * pow(10.00, -5) * elevation)) * extraSolarRadiation
    averageT = (pow(maxTair + 273.160, 4) + pow(minTair + 273.160, 4)) / 2.00
    surfaceEmissivity = 0.340 - (0.140 * sqrt(vaporPressure / 10.00))
    cloudCoverFactor = 1.350 * (solarRadiation / clearSkySolarRadiation) - 0.350
    Nolr = stefanBoltzman * averageT * surfaceEmissivity * cloudCoverFactor
    if ih != -999:
        Nolr = Nolr / 24.00
    netRadiation = Nsr - Nolr
    netOutGoingLongWaveRadiation = Nolr
    return (netRadiation, netOutGoingLongWaveRadiation)
#%%CyML Model End%%